### PR TITLE
Multi-Page Mode via SW

### PR DESCRIPTION
**Pull request by ikreymer**
This sets up a different mode, as suggested in #20, where the rendering of the original urls is taken over by the service workers.
This uses the low-level https://github.com/webrecorder/wabac.js directly instead of replayweb.page and that adds some flexibility.

Since URLs are preserved, suggestions, search and random should work as in the old system.
This system adds several static files that are stored in the `./replay/` folder and also installs sw.js directly from npm.
Indexing and language are also enabled with this. Also, empty payload records are not written to address #21 

The new files in `./replay` are:
- `index.html` - initial page, which will redirect to the main page. Also displays loading message or error message.
- `load.js` - the script used in `index.html`
- `topFrame.html` - the top frame page used on all pages. The banner is injected here by kiwix-serve.
- `sw.js` - the service worker, installed from npm automatically. Loaded from: https://cdn.jsdelivr.net/npm/@webrecorder/wabac/
- `notFoundPage.html` - Generic error page template for when a 404 is encountered. Redirects to live web if not in current domain. Also checks if kiwix-serve is blocking redirects and then redirects to `/catch/external`

This PR addresses:
 - #5 - the wabac.js is bundled directly
 - #17 - random should work
 - #18 - suggestions should work
 - #19 - search should work
 - #20 - no more single page
 - #21 - don't write empty payload A/ records, only H/ records.
 - #22 and #23 - external links should now redirect. if the blocking message is enabled in kiwix-serve it will be displayed.
 - #26 - new loading is just a text message, can be modified in `index.html`

### CODE REVIEW

File: src/warc2zim/replay/index.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<html>
+<head>
```
Comment (rgaudin) [inline src/warc2zim/replay/index.html:2]: You need to set encoding otherwise the taskbar might be displayed incorrectly if ZIM name uses non-ASCII chars.

Comment (rgaudin) [reply src/warc2zim/replay/index.html:2]: Notfound and Top as well

File: src/warc2zim/replay/notFoundPage.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+<script>
+  const URL = "$URL";
+  const mainUrl = "$MAIN_URL";
+
+  if (window.parent === window.top) {
+
+    // check if URL is from same domain as the main url
+    if (!URL.startsWith(mainUrl)) {
+      // check if kiwix-serve 'blocking' is enabled, show interstitional if it is
+      if (window.block_path) {
```
Comment (rgaudin) [inline src/warc2zim/replay/notFoundPage.html:12]: Please submit a PR on kiwix-lib to add a comment on top of this so maintainers know that other tools rely on this
https://github.com/kiwix/kiwix-lib/blob/5f8c0998291d0778887373355bb6b1c5a345ed2c/static/skin/block_external.js#L1

File: src/warc2zim/replay/topFrame.html [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  height: 100%;
+  margin: 0;
+  padding: 0;
+  border: 0;
+  overflow: hidden;
+}
+
+iframe {
+  width: 100%;
+  height: 100%;
```
Comment (rgaudin) [inline src/warc2zim/replay/topFrame.html:16]: Might be a viewport problem but this revived the unreachable content bug ; in FF this time.

File: src/warc2zim/replay/topFrame.html [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  window.addEventListener("message", function() {
+    if (event.data.wb_type === "load" || event.data.wb_type === "replace-url") {
+      
+      // remove scheme to be consistent with current canonicalization
+      const urlNoScheme = event.data.url.slice(event.data.url.indexOf("//") + 2);
+      window.history.replaceState(null, "", PREFIX + urlNoScheme);
+    }
+  });
+
+  iframe.src = PREFIX + "mp_/" + URL;
```
Comment (rgaudin) [inline src/warc2zim/replay/topFrame.html:39]: Is `mp_` a sort of namespace?  Where does this come from? sw.js?

File: src/warc2zim/main.py [hunk 5]
```diff
@@ -21,6 +21,8 @@
 from argparse import ArgumentParser, RawTextHelpFormatter
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply src/warc2zim/main.py:21]: Ok, will remove in subsequent PR

File: src/warc2zim/main.py [hunk 6]
```diff
@@ -155,55 +163,33 @@ def get_data(self):
...
+class RWPStaticArticle(BaseArticle):
     def __init__(self, filename, main_url):
-        super(RWPViewerArticle, self).__init__()
+        super(RWPStaticArticle, self).__init__()
         self.filename = filename
         self.main_url = main_url
 
+        self.mime, _ = mimetypes.guess_type(filename)
+        self.content = pkg_resources.resource_string(
+            "warc2zim", "replay/" + filename
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply src/warc2zim/main.py:174]: This is using `pkg_resources` to load data from the package, which always use `/`, not the file system:
https://setuptools.readthedocs.io/en/latest/pkg_resources.html#basic-resource-access

There may be better ways to do this, will look.

File: src/warc2zim/main.py [hunk 7]
```diff
@@ -224,21 +210,28 @@ def __init__(self, args):
...
-            )
+        self.add_remote_or_local("sw.js")
+
+        for filename in pkg_resources.resource_listdir("warc2zim", "replay"):
+            if filename != "sw.js":
+                self.replay_articles.append(RWPStaticArticle(filename, self.main_url))
 
         with Creator(
-            self.name, main_page="viewer.html", index_language="", min_chunk_size=8192
+            self.name, main_page="index.html", index_language="en"
```
[the remark this answers is not in the document]
Comment (ikreymer) [reply src/warc2zim/main.py:229]: This is fixed in another branch that I'll add after..

File: src/warc2zim/replay/index.html (hunk 1 above)
Comment (ikreymer) [reply src/warc2zim/replay/index.html:2]: Good point! Will fix in follow up.

File: src/warc2zim/replay/topFrame.html (hunk 4 above)
Comment (ikreymer) [reply src/warc2zim/replay/topFrame.html:39]: Yes, sort of, it is a part of the replay.. Probably needs better documentation in wabac.js
I have it documented in pywb, which is the python version of the rewriting system, but it is the same idea:
https://pywb.readthedocs.io/en/latest/manual/rewriter.html#main-page-modifier-mp

Without the modifier, it serves the outer frame.
With the `mp_/` modifier it serves the rewritten content.

**Pull request merged**